OP_CODEGEN_GUIDE.md

```diff
@@ -0,0 +1,209 @@
+- xla/torch_xla/csrc/generated/XLANativeFunctions.cpp
+  - Result of the full_codegen column of the xla/xla_native_functions.yaml. The op function defined here will implement the op declared in the XLANativeFunctions.h. Each op will take at::tensor and return another at::tensor wrapped around a XLATensor.
+- xla/torch_xla/csrc/generated/LazyIr.h
+  - Result of the full_codegen column of the xla/xla_native_functions.yaml.  Defines the IR that is used to construct the full_codegen ops.
+
+### PyTorch/XLA Old Op Lowering files
+- xla/torch_xla/csrc/generated/aten_xla_type.cpp
+  - Manually implements ops defined in xla/xla_native_functions.yaml. Will be replaced by XLANativeFunctions.cpp
+- xla/torch_xla/csrc/generated/tensor.h
+  - Defines XLATensor class and XLATensor method declarations. These declarations are usually a one to one mapping of the at::Tensor nodes we declared in XLANativeFunctions.h. XLATensor method will be removed for full_codegen ops
+- xla/torch_xla/csrc/generated/tensor_method.cpp
+  - Implements tensor methods defined in tensor.h. This file will be removed for full_codegen ops
+- xla/torch_xla/csrc/generated/ops/…
+  - Defines IR class for “most” ops. It is possible that multiple ops share the same IR.
+
+## Codegen step by step
+### 1. Identify the op 
+When you work on your first few codegens, we generally recommend you to start with the simpler ops. This guide will go over one unary one one binary op as examples, but it is recommend that you avoid ops with the following characteristics:
+1. Contains custom fallback code. For example in _adaptive_avg_pool3d, there is a conditional fallback:
+```
+  if (!IsSupportedAdaptivePool(XlaHelpers::I64List(self.sizes()),
+                               output_size_list, /*pool_dim=*/3)) {
+    return at::native::call_fallback_fn<&xla_cpu_fallback, ATEN_OP(_adaptive_avg_pool3d)>::call(self, output_size);
+  }
+```
+2. Results in dynamic shape as these ops are WIP and may evolve over time. At some future point, we may bring the ops into codegen. 
+3. Does not invoke a tensor_method directly. For example _copy_from:
+```
+ if (!self_tensor) {
+   static bool sync_update =
+       xla::sys_util::GetEnvBool("XLA_TENSOR_UPDATE_SYNC", true);
+   XLA_CHECK(dst_tensor);
+   dst_tensor->UpdateFromTensor(self, /*sync=*/sync_update);
+ }
+```
+4. Has a complicated tensor_method, ideally it should be a directly mapping from op to IR.
+
+An good example of a "simple" op would be something like `abs`:
+```
+at::Tensor XLANativeFunctions::abs(const at::Tensor& self) {
+  XLA_FN_COUNTER("xla::");
+  return bridge::AtenFromXlaTensor(XLATensor::abs(bridge::GetXlaTensor(self)));
+}
+```
+
+### 2. Codegen the op and inspect the generated file
+Find the op in  `xla/xla_native_functions.yaml` and move it to the full_codegen column and run the build again. The build will fail (reason explained later in this guide) but you can still see the generated file. The code snippets below uses `abs` as an example.
```

Comment:
instead of `run the build again`, maybe make it more concrate like run `python setup.py install under xla dir`